### ISSUE TITLE
prometheus: Why default to NodeHostName in syncNodeEndpoints()

### ISSUE THREAD

**Issue by mxinden** [coreos/prometheus-operator#1796]
The Prometheus Operator [populates](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/operator.go#L442) the kubelet endpoint inside the `kube-system` namespace in order for Prometheus instances to discover kubelets.

It tries to use the [`NodeInternalIP`, otherwise uses `NodeExternalIP` and last the `NodeHostName`](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/operator.go#L430-L438).

Why do we consider using the `NodeHostName`? An [`EndpointAddress.IP` has to be a valid IP address](https://github.com/kubernetes/kubernetes/blob/8ebc84e71d7eb1c45dc296945dc232e811b2b820/pkg/apis/core/types.go#L3287-L3290). Do some setups set the host name to the IP address?

Prometheus Operator version 0.23.1

**Comment by brancz** [comment]
I think this might have been an artifact of Prometheus or Kubernetes code to select an IP, but a hostname was valid in that context. We should remove it.

https://github.com/prometheus/prometheus/blob/71855a22a456cfd25858ab3840fcdf8d590edd09/discovery/kubernetes/node.go#L199-L218

**Comment by mrsiano** [comment]
what about change the address priority instead ? i.e
1. HostName
2. InternalIP
...

**Comment by vsliouniaev** [comment]
@mrsiano that won't work in cases where DNS doesn't resolve the host name. In this case a fall-through to HostName results in errors like
```
Error from server: Get https://worker-001:10250/containerLogs/monitoring/prometheus-operator-788fb64c46-lsv24/prometheus-operator: dial tcp: lookup worker-001 on 10.10.10.10:53: no such host
```

**Comment by vsliouniaev** [comment]
@mxinden I suspect that a configuration where only Hostnames are returned is valid in some Kubernetes setups. The priority can be altered in apiserver settings

**Comment by brancz** [comment]
@carlpett reported the same behavior in #1815, and figured out this is a fault in the kubelet, as the particular kubelet. I noted https://github.com/coreos/prometheus-operator/issues/1815#issuecomment-414957563 what we should do.

**Comment by brancz** [comment]
This was fixed by #1816
